chromium_src/chrome/browser/ssl/https_upgrades_navigation_throttle.cc

```diff
@@ -0,0 +1,103 @@
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ssl/https_only_mode_controller_client.h"
+#include "components/prefs/pref_service.h"
+#include "components/security_interstitials/content/stateful_ssl_host_state_delegate.h"
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/features.h"
+
+namespace content {
+class BrowserContext;
+}  // namespace content
+
+namespace {
+
+// Tor is slow and needs a longer fallback delay
+constexpr base::TimeDelta kTorFallbackDelay = base::Seconds(20);
+
+bool ShouldUpgradeToHttps(content::NavigationHandle* handle) {
+  if (!brave_shields::IsHttpsByDefaultFeatureEnabled()) {
+    return false;
+  }
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  const GURL& url = handle->GetURL();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+  return brave_shields::ShouldUpgradeToHttps(
+      map, url, g_brave_browser_process->https_upgrade_exceptions_service());
+}
+
+bool IsTor(content::NavigationHandle* handle) {
+  auto* context = handle->GetWebContents()->GetBrowserContext();
+  Profile* profile = Profile::FromBrowserContext(context);
+  return profile->IsTor();
+}
+
+}  // namespace
+
+#define MaybeCreateThrottleFor MaybeCreateThrottleFor_ChromiumImpl
+#define SetNavigationTimeout(DEFAULT_TIMEOUT)                         \
+  SetNavigationTimeout(IsTor(navigation_handle()) ? kTorFallbackDelay \
+                                                  : DEFAULT_TIMEOUT)
+
+#include "src/chrome/browser/ssl/https_upgrades_navigation_throttle.cc"
+
+#undef MaybeCreateThrottleFor
+#undef SetNavigationTimeout
+
+// static
+std::unique_ptr<HttpsUpgradesNavigationThrottle>
+HttpsUpgradesNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* handle,
+    std::unique_ptr<SecurityBlockingPageFactory> blocking_page_factory,
+    PrefService* prefs) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  // HTTPS-First Mode is only relevant for primary main-frame HTTP(S)
+  // navigations.
+  if (!handle->GetURL().SchemeIsHTTPOrHTTPS() ||
+      !handle->IsInPrimaryMainFrame() || handle->IsSameDocument()) {
+    return nullptr;
+  }
+
+  const GURL& request_url = handle->GetURL();
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  HostContentSettingsMap* map =
+      HostContentSettingsMapFactory::GetForProfile(context);
+
+  bool https_first_mode_enabled =
+      (base::FeatureList::IsEnabled(features::kHttpsFirstModeV2) && prefs &&
+       prefs->GetBoolean(prefs::kHttpsOnlyModeEnabled)) ||
+      brave_shields::ShouldForceHttps(map, request_url);
+  bool https_upgrades_enabled =
+      (https_first_mode_enabled ||
+       base::FeatureList::IsEnabled(net::features::kBraveHttpsByDefault)) &&
+      ShouldUpgradeToHttps(handle);
```

Comment:
you already have `map` and `request_url`.

you can call `brave_shields::ShouldUpgradeToHttps` directly, bc it contains `IsHttpsByDefaultFeatureEnabled` check.

Reply:
Great point! Fixed.